### PR TITLE
Make mounting ssl cert file optional

### PR DESCRIPTION
**Pull request by layus**
# Motivation

Fix breakage of master after https://github.com/NixOS/nix/pull/7312 got in.

# Context

See https://github.com/NixOS/nix/pull/7312 and https://github.com/NixOS/nix/pull/8375

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by layus** [comment]
I guess this feature deserves a test. No idea how to do that, but I can investigate.

**Comment by Ericson2314** [comment]
@layus see the test that was failing (various `remote-store.sh`-invoking tests). You can use `--store` to force sandboxing, a fixed output derivation `--ssl-cert-file` to force existent and non-exist paths, and then stat them inside your fixed output derivation.

File: src/libstore/build/local-derivation-goal.cc [hunk 1]
```diff
@@ -1777,7 +1777,7 @@ void LocalDerivationGoal::runChild()
                     if (pathExists(path))
                         ss.push_back(path);
 
-                dirsInChroot.emplace(settings.caFile, "/etc/ssl/certs/ca-certificates.crt");
+                dirsInChroot.try_emplace("/etc/ssl/certs/ca-certificates.crt", settings.caFile, true);
```
Comment (Ericson2314) [inline src/libstore/build/local-derivation-goal.cc:1780]: I know it doesn't change behavior, but I would still do
```suggestion
                if (settings.caFile != "")
                    dirsInChroot.try_emplace("/etc/ssl/certs/ca-certificates.crt", settings.caFile, true);
```
Setting a non-existent file is a dubious thing that should get a log message. Setting the empty string is the preferred way to not set any SSL path. Even once a `std::optional` refactor happens, an empty string would still be the surface syntax for `nix.confg` and the env vars.

**Comment by layus** [comment]
@Ericson2314 Fixed. I am more worried about the arguments order though. I got it wrong on the first attempt, so a test should definitely be added.

**Comment by Ericson2314** [comment]
Oh that can still be done the same way. We don't need to pass real certs, just any old junk for the fixed output derivation to `stat`/`cat`/etc.

**Review by Ericson2314 — COMMENTED** [review]
I am also hesitant about "friendly" ignoring config/arguments that make no sense. (See #7918 for another instance of that same tendency of mine.) But the alternatives (like making a special case for the thing Nixpkgs happens to set today) seem no better, and dubious in terms of backwards compat. And we do need to fix CI pronto.

The least-bad of the other options might be continuing to skip env var (`NIX_SSL_CERT_FILE` / `SSL_CERT_FILE`) if the path doesn't exist, but *not* skipping the other brand new ways of setting the config file. But I am not at all sure that is better than this.

Hoping some other reviewer can take this decision out of my hands :)

**Comment by fogti** [comment]
> I am also hesitant about "friendly" ignoring config/arguments that make no sense.

Yeah. I think a good rule of thumb is: hard error on invalid/unknown config names (could be typos and are a hassle to find usually), warning on values that might indicate empty/default in a weird way (e.g. due to backward compat stuff), different from the normal way to do that, should result in a warning (which should probably, in this case, not just logged to standard output but also to daemon logs, etc. idk), error on values that are not empty, but still invalid.

**Comment by layus** [comment]
I agree of course on the principles, but in this case I think it is okay because the behaviour is consistent with nix as a whole. If NIX_SSL_CERT_FILE is set incorrectly, then native fetchurl will also fail. So this will fail just as silently as nix would. No change. Plus this feature is new and thus unrelied upon and de facto experimental. Considering that it fixes CI, I would not care too much about details of failure modes in this particular case.

Fixing the more general problem is important of course, but even in that case SSL config is tricky to error out. How do you know that the invalid value was not intended. /no-cert-file.pem is certainly intended, and means that you get nothing to use. The semantics for all the applications that use SSL is clear, and nix should do the same: work as if no certs are available.

tl;dr; the behaviour sucks, but is the one gerenally acknowledged for SSL_CERT* env vars.

Le 20 mai 2023 01:05:48 GMT+02:00, Alain Zscheile ***@***.***> a écrit :

**Comment by layus** [comment]
I have spent way too much time trying to figure out how to make this behave in a nicer way, and I do not see how to warn the user without getting in the way of legit uses. The problem is that passing a nonexistent SSL_CERT_FILE seems to be the usual way to tell apps not to use the default cert locations. Passing "" to SSL_CERT_FILE env var is often considered just as an unset env var, probably because it is easier to unset env vars by setting them to empty than properly unregistering them from the env. Probably a confusion coming from bash. So you have to support nonexistent paths without barfing at the user. This makes it impossible to warn at mistakes with missing paths intended to exist.

**Comment by Ericson2314** [comment]
> I have spent way too much time trying to figure out how to make this behave in a nicer way, and I do not see how to warn the user without getting in the way of legit uses.

Oh sorry @layus, nothing I was saying was me asking you to do more work. I am just saying I want it wait for someone other person with commit bit to weigh in here before merging.

I am 70% sure they will tell me I'm being a weeny and what you've done is fine :). Then we'll merge it.

**Comment by fogti** [comment]
> The problem is that passing a nonexistent SSL_CERT_FILE seems to be the usual way to tell apps not to use the default cert locations.

oh, ok then.

**Comment by layus** [comment]
Now with tests :-).

> I have spent way too much time trying to figure out how to make this behave in a nicer way, and I do not see how to warn the user without getting in the way of legit uses.

I failed to convey that I did this to myself without any external nudge. It's the way I usually work. I spend too much time on not-so important things just to ensure I get to the bottom of it. So not targeted at anyone in particular but me :-).

File: tests/linux-sandbox-cert-test.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{ fixed-output }:
```
Comment (edolstra) [inline tests/linux-sandbox-cert-test.nix:1]: This test is not called anywhere.

(However VM tests are pretty expensive, so maybe it's overkill to have a separate test for this...)

File: src/libstore/build/local-derivation-goal.cc [hunk 3]
```diff
@@ -1777,7 +1777,8 @@ void LocalDerivationGoal::runChild()
                     if (pathExists(path))
                         ss.push_back(path);
 
-                dirsInChroot.emplace(settings.caFile, "/etc/ssl/certs/ca-certificates.crt");
+                if (settings.caFile != "")
+                    dirsInChroot.try_emplace("/etc/ssl/certs/ca-certificates.crt", settings.caFile, true);
```
Comment (thufschmitt) [inline src/libstore/build/local-derivation-goal.cc:1781]: (for a follow-up): This syntax is quite unfortunate, we could use designated initializers to make it more understandable

Comment (thufschmitt) [reply src/libstore/build/local-derivation-goal.cc:1781]: Also for later: Maybe we should warn if the file doesn't exist

File: tests/linux-sandbox.sh [hunk 4]
```diff
@@ -40,3 +40,27 @@ grepQuiet 'may not be deterministic' $TEST_ROOT/log
 
 # Test that sandboxed builds cannot write to /etc easily
 (! nix-build -E 'with import ./config.nix; mkDerivation { name = "etc-write"; buildCommand = "echo > /etc/test"; }' --no-out-link --sandbox-paths /nix/store)
+
+
+## Test mounting of SSL certificates into the sandbox
+testCert () {
+    (! nix-build linux-sandbox-cert-test.nix --argstr fixed-output "$2" --no-out-link --sandbox-paths /nix/store --option ssl-cert-file "$3" 2> $TEST_ROOT/log)
+    cat $TEST_ROOT/log
+    grepQuiet "CERT_${1}_IN_SANDBOX" $TEST_ROOT/log
```
Comment (Ericson2314) [inline tests/linux-sandbox.sh:49]:
```suggestion
    expectStderr ??? nix-build linux-sandbox-cert-test.nix --argstr fixed-output "$2" --no-out-link --sandbox-paths /nix/store --option ssl-cert-file "$3" 2> $TEST_ROOT/log) \
        | tee /dev/stderr \
        | grepQuiet "CERT_${1}_IN_SANDBOX" $TEST_ROOT/log
```

**Review by thufschmitt — APPROVED** [review]

File: tests/linux-sandbox.sh (hunk 4 above)
Comment (Ericson2314) [reply tests/linux-sandbox.sh:49]: And all the other `(!` too.

Comment (fogti) [reply tests/linux-sandbox.sh:49]: why does the closing `)` stay, but not the opening one?
